File: trunk/dotnet/picocontainer/src/Test/Defaults/TransientComponentAdapterTestCase.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using System.Diagnostics;

using NUnit.Framework;

using PicoContainer.Defaults;
using PicoContainer.Tests.Tck;
using PicoContainer.Tests.TestModel;

namespace PicoContainer.Tests.Defaults
{
	
  [TestFixture]
	public class TransientComponentAdapterTestCase
	{
		public virtual void  testNonCachingComponentAdapterReturnsNewInstanceOnEachCallToGetComponentInstance()
		{
			TransientComponentAdapter componentAdapter = new TransientComponentAdapter(null, typeof(object));
			object o1 = componentAdapter.GetComponentInstance(null);
			object o2 = componentAdapter.GetComponentInstance(null);
			Assert.IsTrue(o1 != o2);
		}
		
		public class Service
		{
		}
		
		public class TransientComponent
		{
			public Service service;
			
			public TransientComponent(Service service)
			{
				this.service = service;
			}
		}
		
		public virtual void  testDefaultPicoContainerReturnsNewInstanceForEachCallWhenUsingTransientComponentAdapter()
		{
			DefaultPicoContainer picoContainer = new DefaultPicoContainer();
			picoContainer.RegisterComponentImplementation(typeof(Service));
			picoContainer.RegisterComponent(new TransientComponentAdapter(typeof(TransientComponent), typeof(TransientComponent)));
			TransientComponent c1 = (TransientComponent) picoContainer.GetComponentInstance(typeof(TransientComponent));
			TransientComponent c2 = (TransientComponent) picoContainer.GetComponentInstance(typeof(TransientComponent));
			Assert.IsTrue(c1 != c2);
			Assert.AreEqual(c1.service, c2.service);
		}
	}
}

File: trunk/java/picocontainer/src/test/org/picocontainer/defaults/TransientComponentAdapterTestCase.java
package org.picocontainer.defaults;

import junit.framework.TestCase;
import org.picocontainer.ComponentAdapter;

public class TransientComponentAdapterTestCase extends TestCase {
    public void testNonCachingComponentAdapterReturnsNewInstanceOnEachCallToGetComponentInstance() {
        TransientComponentAdapter componentAdapter = new TransientComponentAdapter(null, Object.class);
        Object o1 = componentAdapter.getComponentInstance(null);
        Object o2 = componentAdapter.getComponentInstance(null);
        assertNotSame(o1, o2);
    }

    public static class Service {
    }

    public static class TransientComponent {
        private Service service;

        public TransientComponent(Service service) {
            this.service = service;
        }
    }

    public void testDefaultPicoContainerReturnsNewInstanceForEachCallWhenUsingTransientComponentAdapter() {
        DefaultPicoContainer picoContainer = new DefaultPicoContainer();
        picoContainer.registerComponentImplementation(Service.class);
        picoContainer.registerComponent(new TransientComponentAdapter(TransientComponent.class, TransientComponent.class));
        TransientComponent c1 = (TransientComponent) picoContainer.getComponentInstance(TransientComponent.class);
        TransientComponent c2 = (TransientComponent) picoContainer.getComponentInstance(TransientComponent.class);
        assertNotSame(c1, c2);
        assertSame(c1.service, c2.service);
    }


    public static class A {
        public A() {
            fail("verification should not instantiate");
        }
    }

    public static class B {
        public B(A a) {
            fail("verification should not instantiate");
        }
    }

    public void testSuccessfulVerificationWithNoDependencies() {
        TransientComponentAdapter componentAdapter = new TransientComponentAdapter("foo", A.class);
        componentAdapter.verify(null);
    }

    public void testFailingVerificationWithUnsatisfiedDependencies() {
        ComponentAdapter componentAdapter = new TransientComponentAdapter("foo", B.class);
        try {
            componentAdapter.verify(new DefaultPicoContainer());
            fail("Expected NoSatisfiableConstructorsException");
        } catch (NoSatisfiableConstructorsException e) {
            // ok
        }
    }

}
